File: libs/eave-stdlib-py/src/eave/stdlib/task_queue.py
import asyncio
from dataclasses import dataclass
import json
from typing import Any, Coroutine, Optional, TypeVar
from google.cloud import tasks
from starlette.requests import Request
from eave.stdlib import cache
import eave.stdlib.signing as signing
from eave.stdlib.eave_origins import EaveApp
from eave.stdlib.headers import (
    CONTENT_TYPE,
    EAVE_ACCOUNT_ID_HEADER,
    EAVE_ORIGIN_HEADER,
    EAVE_REQUEST_ID_HEADER,
    EAVE_SIG_TS_HEADER,
    EAVE_SIGNATURE_HEADER,
    EAVE_TEAM_ID_HEADER,
    GCP_CLOUD_TRACE_CONTEXT,
    GCP_GAE_REQUEST_LOG_ID,
    USER_AGENT,
)
from eave.stdlib.time import ONE_DAY_IN_MS
from eave.stdlib.util import compact_deterministic_json, ensure_bytes, ensure_str

from .typing import JsonObject
from .config import shared_config
from .logging import LogContext, eaveLogger

T = TypeVar("T")

asyncio_tasks = set[asyncio.Task[Any]]()


def do_in_background(coro: Coroutine[Any, Any, T]) -> asyncio.Task[T]:
    task = asyncio.create_task(coro)
    asyncio_tasks.add(task)
    task.add_done_callback(asyncio_tasks.discard)
    return task


async def get_queue(queue_name: str) -> tasks.Queue:
    client = tasks.CloudTasksAsyncClient()

    queue = client.queue_path(
        project=shared_config.google_cloud_project,
        location=shared_config.app_location,
        queue=queue_name,
    )

    queue = await client.get_queue(name=queue)
    return queue


@dataclass
class BodyCacheEntry:
    cache_key: str


async def create_task_from_request(
    queue_name: str,
    target_path: str,
    request: Request,
    origin: EaveApp,
    audience: EaveApp,
    ctx: Optional[LogContext],
    unique_task_id: Optional[str] = None,
    task_name_prefix: Optional[str] = None,
) -> None:
    ctx = LogContext.wrap(ctx)

    if not unique_task_id:
        if trace_id := request.headers.get(GCP_CLOUD_TRACE_CONTEXT):
            unique_task_id = trace_id.split("/")[0]
        elif log_id := request.headers.get(GCP_GAE_REQUEST_LOG_ID):
            unique_task_id = log_id

    # Stash in redis to avoid 100kb payload size limit in Cloud Tasks
    payload = await request.body()
    cache_client = cache.client_or_exception()
    pointer_payload = BodyCacheEntry(cache_key=ctx.eave_request_id)
    await cache_client.set(name=pointer_payload.cache_key, value=ensure_str(payload), ex=ONE_DAY_IN_MS)
    payload = compact_deterministic_json(pointer_payload.__dict__)

    headers = dict(request.headers)

    # The "user agent" is Slack Bot when coming from Slack, but for the task processor that's not the case.
    headers.pop(USER_AGENT, None)

    await create_task(
        queue_name=queue_name,
        target_path=target_path,
        payload=payload,
        origin=origin,
        audience=audience,
        unique_task_id=unique_task_id,
        task_name_prefix=task_name_prefix,
        headers=headers,
        ctx=ctx,
    )


async def create_task(
    queue_name: str,
    target_path: str,
    payload: JsonObject | str | bytes,
    origin: EaveApp,
    audience: EaveApp,
    ctx: Optional[LogContext],
    unique_task_id: Optional[str] = None,
    task_name_prefix: Optional[str] = None,
    headers: Optional[dict[str, str]] = None,
) -> tasks.Task:
    ctx = LogContext.wrap(ctx)

    if not headers:
        headers = {}

    eave_sig_ts = signing.make_sig_ts()

    team_id = headers.get(EAVE_TEAM_ID_HEADER) or ctx.eave_team_id
    account_id = headers.get(EAVE_ACCOUNT_ID_HEADER) or ctx.eave_account_id
    request_id = headers.get(EAVE_REQUEST_ID_HEADER) or ctx.eave_request_id

    signature_message = signing.build_message_to_sign(
        method="POST",
        origin=origin,
        ts=eave_sig_ts,
        audience=audience,
        request_id=request_id,
        path=target_path,
        payload=ensure_str(payload),
        team_id=team_id,
        account_id=account_id,
        ctx=ctx,
    )

    signature = signing.sign_b64(signing_key=signing.get_key(origin), data=signature_message)

    headers[CONTENT_TYPE] = "application/json"
    headers[EAVE_SIGNATURE_HEADER] = signature
    headers[EAVE_SIG_TS_HEADER] = str(eave_sig_ts)
    headers[EAVE_ORIGIN_HEADER] = origin.value

    if account_id and not headers.get(EAVE_ACCOUNT_ID_HEADER):
        headers[EAVE_ACCOUNT_ID_HEADER] = account_id
    if team_id and not headers.get(EAVE_TEAM_ID_HEADER):
        headers[EAVE_TEAM_ID_HEADER] = team_id
    if request_id and not headers.get(EAVE_REQUEST_ID_HEADER):
        headers[EAVE_REQUEST_ID_HEADER] = request_id

    client = tasks.CloudTasksAsyncClient()

    parent = client.queue_path(
        project=shared_config.google_cloud_project,
        location=shared_config.app_location,
        queue=queue_name,
    )

    task = tasks.Task(
        app_engine_http_request=tasks.AppEngineHttpRequest(
            http_method=tasks.HttpMethod.POST,
            relative_uri=target_path,
            headers=headers,
            body=ensure_bytes(data=payload),
        )
    )

    if unique_task_id:
        if task_name_prefix:
            unique_task_id = f"{task_name_prefix}{unique_task_id}"

        # If this isn't given, Cloud Tasks creates a unique task name automatically.
        task.name = client.task_path(
            project=shared_config.google_cloud_project,
            location=shared_config.app_location,
            queue=queue_name,
            task=unique_task_id,
        )

    eaveLogger.debug(
        f"Creating task on queue {queue_name}",
        ctx,
        {
            "task_name": task.name,
            "queue_name": parent,
        },
    )

    t = await client.create_task(parent=parent, task=task)
    return t


async def get_cached_payload(body: str | bytes) -> str:
    jbody = json.loads(body)
    try:
        pointer_payload = BodyCacheEntry(**jbody)
    except TypeError as e:
        eaveLogger.error("Invalid BodyCacheEntry payload!")
        raise e

    cache_client = cache.client_or_exception()
    stashed_payload = await cache_client.get(name=pointer_payload.cache_key)
    assert stashed_payload is not None, "Could not find expected cached event body. Maybe the TTL needs to be extended?"

    return stashed_payload

File: libs/eave-stdlib-ts/src/task-queue.ts
import { CloudTasksClient, protos } from "@google-cloud/tasks";
import { Request } from "express";
import assert from "node:assert";
import { constants as httpConstants } from "node:http2";
import { ClientApiEndpointConfiguration } from "./api-util.js";
import { getCacheClient } from "./cache.js";
import { sharedConfig } from "./config.js";
import { EaveApp } from "./eave-origins.js";
import {
  EAVE_ACCOUNT_ID_HEADER,
  EAVE_ORIGIN_HEADER,
  EAVE_REQUEST_ID_HEADER,
  EAVE_SIGNATURE_HEADER,
  EAVE_SIG_TS_HEADER,
  EAVE_TEAM_ID_HEADER,
  GCP_CLOUD_TRACE_CONTEXT,
  GCP_GAE_REQUEST_LOG_ID,
  MIME_TYPE_JSON,
} from "./headers.js";
import { LogContext, eaveLogger } from "./logging.js";
import { CtxArg, makeRequest } from "./requests.js";
import Signing, { buildMessageToSign, makeSigTs } from "./signing.js";
import { ExpressRoutingMethod } from "./types.js";
import { makeString } from "./util.js";

type CreateTaskSharedArgs = CtxArg & {
  queueName: string;
  targetPath: string;
  origin: EaveApp;
  audience: EaveApp;
  uniqueTaskId?: string;
  taskNamePrefix?: string;
};

type CreateTaskArgs = CreateTaskSharedArgs & {
  payload: any;
  headers?: { [key: string]: string };
};

type CreateTaskFromRequestArgs = CreateTaskSharedArgs & {
  req: Request;
};

type BodyCacheEntry = {
  cacheKey: string;
};

export async function createTaskFromRequest({
  queueName,
  targetPath,
  origin,
  audience,
  uniqueTaskId,
  taskNamePrefix,
  req,
  ctx,
}: CreateTaskFromRequestArgs): Promise<void> {
  ctx = LogContext.wrap(ctx, req);

  // keep the event in redis since GCP Task Queue can only accept 100kb task size
  let payload = req.body;
  const pointerPayload: BodyCacheEntry = { cacheKey: ctx.eave_request_id };
  const cacheClient = await getCacheClient();
  await cacheClient.set(pointerPayload.cacheKey, makeString(payload), {
    EX: 60 * 60 * 24,
  }); // TTL 24h

  payload = makeString(pointerPayload);

  if (!uniqueTaskId) {
    const traceId = req.header(GCP_CLOUD_TRACE_CONTEXT);
    const logId = req.header(GCP_GAE_REQUEST_LOG_ID);

    if (traceId) {
      uniqueTaskId = traceId.split("/")[0];
    } else if (logId) {
      uniqueTaskId = logId;
    }
  }

  const headers: { [key: string]: string } = {};
  // FIXME: Is there a cleaner way to do this? req.headers is a NodeJS.Dict typescript object.
  // eslint-disable-next-line no-restricted-syntax
  for (const key in req.headers) {
    if (Object.hasOwn(req.headers, key)) {
      // express joins array values into string values for headers
      const value = req.header(key)!;
      headers[key] = value;
    }
  }

  // The "user agent" is the GitHub webhook deliverer, but when passing off to the task processor, that's not true.
  // GCP Task Queue merges the user agents.
  delete headers[httpConstants.HTTP2_HEADER_USER_AGENT];

  await createTask({
    targetPath,
    queueName,
    audience,
    origin,
    uniqueTaskId,
    taskNamePrefix,
    payload,
    headers,
    ctx,
  });
}

export async function createTask({
  queueName,
  targetPath,
  payload,
  headers,
  origin,
  audience,
  uniqueTaskId,
  taskNamePrefix,
  ctx,
}: CreateTaskArgs): Promise<void> {
  ctx = LogContext.wrap(ctx);

  payload = makeString(payload);

  if (!headers) {
    headers = {};
  }

  const teamId = headers[EAVE_TEAM_ID_HEADER] || ctx.eave_team_id;
  const accountId = headers[EAVE_ACCOUNT_ID_HEADER] || ctx.eave_account_id;
  const requestId = headers[EAVE_REQUEST_ID_HEADER] || ctx.eave_request_id;

  const eaveSigTs = makeSigTs();

  const signatureMessage = buildMessageToSign({
    method: httpConstants.HTTP2_METHOD_POST,
    path: targetPath,
    ts: eaveSigTs,
    requestId,
    origin,
    audience,
    payload,
    teamId,
    accountId,
    ctx,
  });

  const signing = Signing.new(origin);
  const signature = await signing.signBase64(signatureMessage);

  headers[httpConstants.HTTP2_HEADER_CONTENT_TYPE] = MIME_TYPE_JSON;
  headers[EAVE_SIGNATURE_HEADER] = signature;
  headers[EAVE_ORIGIN_HEADER] = origin;
  headers[EAVE_SIG_TS_HEADER] = eaveSigTs.toString();

  if (requestId && !headers[EAVE_REQUEST_ID_HEADER]) {
    headers[EAVE_REQUEST_ID_HEADER] = requestId;
  }

  if (accountId && !headers[EAVE_ACCOUNT_ID_HEADER]) {
    headers[EAVE_ACCOUNT_ID_HEADER] = accountId;
  }

  if (teamId && !headers[EAVE_TEAM_ID_HEADER]) {
    headers[EAVE_TEAM_ID_HEADER] = teamId;
  }

  const client = new CloudTasksClient();
  const parent = client.queuePath(
    sharedConfig.googleCloudProject,
    sharedConfig.appLocation,
    queueName,
  );

  const task: protos.google.cloud.tasks.v2.ITask = {
    appEngineHttpRequest: {
      headers,
      httpMethod: "POST",
      relativeUri: targetPath,
      body: Buffer.from(payload).toString("base64"),
    },
  };

  assert(task.appEngineHttpRequest);

  if (uniqueTaskId) {
    if (taskNamePrefix) {
      uniqueTaskId = `${taskNamePrefix}${uniqueTaskId}`;
    }

    // If this isn't given, Cloud Tasks creates a unique task name automatically.
    task.name = client.taskPath(
      sharedConfig.googleCloudProject,
      sharedConfig.appLocation,
      queueName,
      uniqueTaskId,
    );
  }

  eaveLogger.debug(`Creating task on queue ${queueName}`, ctx, {
    // fields are snake cased for consistency with Python
    task: {
      name: task.name,
      body: task.appEngineHttpRequest.body?.toString(),
      headers: task.appEngineHttpRequest.headers,
    },
    queue_name: parent,
  });

  if (sharedConfig.isDevelopment) {
    const host = sharedConfig.eavePublicServiceBase(origin);

    const endpointConfig: ClientApiEndpointConfiguration = {
      path: task.appEngineHttpRequest.relativeUri!,
      url: `${host}${task.appEngineHttpRequest.relativeUri}`,
      audience,
      method: ExpressRoutingMethod.post,
    };

    /*
    In development only, signature is done twice: once in this function, and again in `makeRequest`. Not ideal but not worth refactoring this function to prevent it.

    The purpose of having this development block at the bottom of the function is to run as much code as possible in development. Cloud Tasks doesn't have an official local emulator as of this writing, so placing a job on the queue isn't useful during development.

    So in development, we bypass the queue and make the request directly to the target app. It's not a great solution! [This](https://github.com/aertje/cloud-tasks-emulator) exists, but it's a) third-party, and b) requires either `go` or `docker`, which currently would complicate development/onboarding too much to be worth it.
    */
    await makeRequest({
      config: endpointConfig,
      origin,
      ctx,
      input: payload, // not b64 encoded
      addlHeaders: task.appEngineHttpRequest.headers || undefined,
    });
  }

  await client.createTask({ parent, task });
}

export async function getCachedPayload(req: Request): Promise<string> {
  const cacheEntry: BodyCacheEntry = req.body;
  if (cacheEntry.cacheKey === undefined) {
    throw Error("Request body did not contain a cache key");
  }

  const cacheClient = await getCacheClient();
  const cachedBody = await cacheClient.get(cacheEntry.cacheKey);
  if (!cachedBody) {
    throw Error(
      "Could not find expected cached event body. Maybe the TTL needs to be extended?",
    );
  }
  return makeString(cachedBody);
}
